File: src-tauri/src/commands/isp_packages.rs
use crate::models::{
    CreateIspPackageRequest, IspPackage, IspPackageRouterMapping, IspPackageRouterMappingView,
    PaginatedResponse, UpdateIspPackageRequest, UpsertIspPackageRouterMappingRequest,
};
use crate::services::{AuthService, IspPackageService};
use tauri::State;

async fn require_isp_package_permission(
    auth: &AuthService,
    claims: &crate::services::auth_service::Claims,
    tenant_id: &str,
    action: &str,
) -> Result<(), String> {
    auth.check_permission(&claims.sub, tenant_id, "isp_packages", action)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_isp_packages(
    token: String,
    q: Option<String>,
    page: Option<u32>,
    per_page: Option<u32>,
    sort_by: Option<String>,
    sort_dir: Option<String>,
    auth: State<'_, AuthService>,
    svc: State<'_, IspPackageService>,
) -> Result<PaginatedResponse<IspPackage>, String> {
    let claims = auth
        .validate_token(&token)
        .await
        .map_err(|e| e.to_string())?;
    let tenant_id = claims
        .tenant_id
        .clone()
        .ok_or_else(|| "No tenant ID in token".to_string())?;
    require_isp_package_permission(&auth, &claims, &tenant_id, "read").await?;

    svc.list_packages(
        &claims.sub,
        &tenant_id,
        q,
        page.unwrap_or(1),
        per_page.unwrap_or(25),
        sort_by,
        sort_dir,
    )
    .await
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_isp_package(
    token: String,
    service_type: Option<String>,
    name: String,
    description: Option<String>,
    features: Option<Vec<String>>,
    is_active: Option<bool>,
    price_monthly: Option<f64>,
    price_yearly: Option<f64>,
    auth: State<'_, AuthService>,
    svc: State<'_, IspPackageService>,
) -> Result<IspPackage, String> {
    let claims = auth
        .validate_token(&token)
        .await
        .map_err(|e| e.to_string())?;
    let tenant_id = claims
        .tenant_id
        .clone()
        .ok_or_else(|| "No tenant ID in token".to_string())?;
    require_isp_package_permission(&auth, &claims, &tenant_id, "manage").await?;

    let dto = CreateIspPackageRequest {
        service_type,
        name,
        description,
        features,
        is_active,
        price_monthly,
        price_yearly,
    };
    svc.create_package(&claims.sub, &tenant_id, dto, Some("127.0.0.1"))
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_isp_package(
    token: String,
    id: String,
    service_type: Option<String>,
    name: Option<String>,
    description: Option<String>,
    features: Option<Vec<String>>,
    is_active: Option<bool>,
    price_monthly: Option<f64>,
    price_yearly: Option<f64>,
    auth: State<'_, AuthService>,
    svc: State<'_, IspPackageService>,
) -> Result<IspPackage, String> {
    let claims = auth
        .validate_token(&token)
        .await
        .map_err(|e| e.to_string())?;
    let tenant_id = claims
        .tenant_id
        .clone()
        .ok_or_else(|| "No tenant ID in token".to_string())?;
    require_isp_package_permission(&auth, &claims, &tenant_id, "manage").await?;

    let dto = UpdateIspPackageRequest {
        service_type,
        name,
        description,
        features,
        is_active,
        price_monthly,
        price_yearly,
    };

    svc.update_package(&claims.sub, &tenant_id, &id, dto, Some("127.0.0.1"))
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_isp_package(
    token: String,
    id: String,
    auth: State<'_, AuthService>,
    svc: State<'_, IspPackageService>,
) -> Result<(), String> {
    let claims = auth
        .validate_token(&token)
        .await
        .map_err(|e| e.to_string())?;
    let tenant_id = claims
        .tenant_id
        .clone()
        .ok_or_else(|| "No tenant ID in token".to_string())?;
    require_isp_package_permission(&auth, &claims, &tenant_id, "manage").await?;

    svc.delete_package(&claims.sub, &tenant_id, &id, Some("127.0.0.1"))
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_isp_package_router_mappings(
    token: String,
    router_id: Option<String>,
    auth: State<'_, AuthService>,
    svc: State<'_, IspPackageService>,
) -> Result<Vec<IspPackageRouterMappingView>, String> {
    let claims = auth
        .validate_token(&token)
        .await
        .map_err(|e| e.to_string())?;
    let tenant_id = claims
        .tenant_id
        .clone()
        .ok_or_else(|| "No tenant ID in token".to_string())?;
    require_isp_package_permission(&auth, &claims, &tenant_id, "read").await?;

    svc.list_router_mappings(&claims.sub, &tenant_id, router_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn upsert_isp_package_router_mapping(
    token: String,
    router_id: String,
    package_id: String,
    router_profile_name: String,
    address_pool: Option<String>,
    auth: State<'_, AuthService>,
    svc: State<'_, IspPackageService>,
) -> Result<IspPackageRouterMapping, String> {
    let claims = auth
        .validate_token(&token)
        .await
        .map_err(|e| e.to_string())?;
    let tenant_id = claims
        .tenant_id
        .clone()
        .ok_or_else(|| "No tenant ID in token".to_string())?;
    require_isp_package_permission(&auth, &claims, &tenant_id, "manage").await?;

    let dto = UpsertIspPackageRouterMappingRequest {
        router_id,
        package_id,
        router_profile_name,
        address_pool,
    };

    svc.upsert_router_mapping(&claims.sub, &tenant_id, dto, Some("127.0.0.1"))
        .await
        .map_err(|e| e.to_string())
}

File: src/lib/api/ispPackages.ts
import { getTokenOrThrow, safeInvoke } from './core';
import type { IspPackage, IspPackageRouterMappingView, PaginatedResponse } from './types';

export const ispPackages = {
  packages: {
    list: (params?: {
      q?: string;
      page?: number;
      per_page?: number;
      sort_by?: string;
      sort_dir?: 'asc' | 'desc';
    }): Promise<PaginatedResponse<IspPackage>> =>
      safeInvoke('list_isp_packages', { token: getTokenOrThrow(), ...(params || {}) }),

    create: (dto: {
      service_type?: string;
      name: string;
      description?: string | null;
      features?: string[];
      is_active?: boolean;
      price_monthly?: number;
      price_yearly?: number;
    }): Promise<IspPackage> =>
      safeInvoke('create_isp_package', {
        token: getTokenOrThrow(),
        service_type: dto.service_type ?? 'internet_pppoe',
        name: dto.name,
        description: dto.description ?? null,
        features: dto.features ?? [],
        is_active: dto.is_active ?? true,
        price_monthly: dto.price_monthly ?? 0,
        price_yearly: dto.price_yearly ?? 0,
      }),

    update: (
      id: string,
      dto: {
        service_type?: string;
        name?: string;
        description?: string | null;
        features?: string[];
        is_active?: boolean;
        price_monthly?: number;
        price_yearly?: number;
      },
    ): Promise<IspPackage> =>
      safeInvoke('update_isp_package', {
        token: getTokenOrThrow(),
        id,
        service_type: dto.service_type ?? undefined,
        name: dto.name,
        description: dto.description ?? undefined,
        features: dto.features,
        is_active: dto.is_active,
        price_monthly: dto.price_monthly,
        price_yearly: dto.price_yearly,
      }),

    delete: (id: string): Promise<void> =>
      safeInvoke('delete_isp_package', { token: getTokenOrThrow(), id }),
  },

  routerMappings: {
    list: (params?: { router_id?: string }): Promise<IspPackageRouterMappingView[]> =>
      safeInvoke('list_isp_package_router_mappings', {
        token: getTokenOrThrow(),
        router_id: params?.router_id,
      }),

    upsert: (dto: {
      router_id: string;
      package_id: string;
      router_profile_name: string;
      address_pool?: string | null;
    }): Promise<any> =>
      safeInvoke('upsert_isp_package_router_mapping', {
        token: getTokenOrThrow(),
        router_id: dto.router_id,
        package_id: dto.package_id,
        router_profile_name: dto.router_profile_name,
        address_pool: dto.address_pool ?? null,
      }),
  },
};
